### PR TITLE
pkg/tableroute: add table route pkg

### PR DESCRIPTION
**Pull request by GregoryIan**
@iamxy PTAL

### CODE REVIEW

**Comment by holys** [comment]
LGTM

File: pkg/tableroute/trieroute.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: 2017

File: pkg/tableroute/trieroute.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	Insert(pattern string, target string) error
+	// Match will match all items that matched to the origin
+	Match(origin string) string
+	// Remove will remove the matched rule
+	Remove(pattern string) error
+	// AllRules will returns all rules
+	AllRules() map[string]string
+}
+
+type trieRouter struct {
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: add wikipedia link for this type of data structure

File: pkg/tableroute/trieroute.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	next    *node
+	literal string
+}
+
+func newNode() *node {
+	return &node{characters: make(map[byte]*item)}
+}
+
+// NewTrieRouter returns a trie Router
+func NewTrieRouter() Router {
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: New()

File: pkg/tableroute/trieroute.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+	return &trieRouter{cache: make(map[string]string), root: newNode(), isEmpty: true}
+}
+
+// Insert implements Router's Insert()
+func (t *trieRouter) Insert(pattern, target string) error {
+	if len(pattern) == 0 || len(target) == 0 {
+		return errors.Errorf("pattern %s and target %s can't be empty", pattern, target)
+	}
+	if pattern[0] == asterisk {
+		return errors.Errorf("invalid pattern %s", pattern)
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: unsupported pattern

File: pkg/tableroute/trieroute.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+	}
+
+	t.Lock()
+
+	n := t.root
+	hadAsterisk := false
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: has

File: pkg/tableroute/trieroute.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+)
+
+// 1. asterisk character (*, also called "star") matches zero or more characters,
+//    for example, doc* matches doc and document but not dodo;
+//    asterisk character must be in the end of wildcard word,
+//    and there is only one asterisk in one wildcard word
+// 2. the question mark ? matches exactly one character
+const (
+	// asterisk [ * ]
+	asterisk = '*'
```
Comment (andelf) [inline pkg/tableroute/trieroute.go]: upper case

Comment (holys) [inline pkg/tableroute/trieroute.go]: @andelf  Why upper case?

Comment (andelf) [inline pkg/tableroute/trieroute.go]: const

File: pkg/tableroute/trieroute.go (hunk 3 above)
Comment (GregoryIan) [inline pkg/tableroute/trieroute.go]: as the interface defined above, maybe there are other implements

**Comment by GregoryIan** [comment]
update: the route rule in syncer would be like
```
schema = "schema_*"
table = "abc_*"
pattern-schema = "schema"
pattern-table = "abc"
```
this rule would route [schema_1, abc_1] to [schema, abc] .
the detail would be in next pr(synced supports table route)
@iamxy @holys PTAL

File: pkg/tableroute/trieroute_test.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+var _ = Suite(&testRouteSuite{})
+
+type testRouteSuite struct{}
+
+func (t *testRouteSuite) TestRoute(c *C) {
+	r := NewTrieRouter()
+	t.testInsert(c, r)
+	//t.testMatch(c, r)
```
Comment (holys) [inline pkg/tableroute/trieroute_test.go]: Why comment here?

File: pkg/tableroute/trieroute.go (hunk 6 above)
Comment (hicqu) [inline pkg/tableroute/trieroute.go]: private variables shoud be lower case?

**Comment by hicqu** [comment]
LGTM.

**Comment by holys** [comment]
LGTM

**Pull request merged**